modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/MvPartitionStorages.java

```diff
@@ -0,0 +1,495 @@
+     */
+    public CompletableFuture<MvPartitionStorage> create(int partitionId, IntFunction<T> createStorageFunction) {
+        StorageOperation storageOperation = operationByPartitionId.compute(partitionId, (partId, operation) -> {
+            if (operation instanceof DestroyStorageOperation) {
+                if (!((DestroyStorageOperation) operation).setCreationOperation(new CreateStorageOperation())) {
+                    throw new StorageException(
+                            "Creation of the storage after its destruction is already planned: [" + createStorageInfo(partitionId) + ']'
+                    );
+                }
+
+                return operation;
+            }
+
+            if (get(partitionId) != null) {
+                throw new StorageException("Storage already exists: [" + createStorageInfo(partitionId) + ']');
+            }
+
+            throwExceptionDependingOnOperation(operation, partitionId);
+
+            return new CreateStorageOperation();
+        });
+
+        CompletableFuture<Void> destroyStorageFuture = storageOperation instanceof DestroyStorageOperation
+                ? ((DestroyStorageOperation) storageOperation).getDestroyFuture()
+                : completedFuture(null);
+
+        return destroyStorageFuture.thenApply(unused -> {
+            T newStorage = createStorageFunction.apply(partitionId);
+
+            boolean set = storageByPartitionId.compareAndSet(partitionId, null, newStorage);
+
+            assert set : createStorageInfo(partitionId);
+
+            return (MvPartitionStorage) newStorage;
+        }).whenComplete((storage, throwable) -> operationByPartitionId.compute(partitionId, (partId, operation) -> {
+            assert operation instanceof CreateStorageOperation : createStorageInfo(partitionId) + ", op=" + operation;
+
+            return null;
+        }));
+    }
+
+    /**
+     * Destroys a multi-versioned partition storage.
+     *
+     * @param partitionId Partition ID.
+     * @param destroyStorageFunction Partition destruction function.
+     * @return Future of multi-versioned partition storage destruction.
+     * @throws IllegalArgumentException If partition ID is out of configured bounds.
+     * @throws StorageException If the storage does not exist or another operation is already in progress.
+     * @throws StorageRebalanceException If the storage is in the process of rebalancing.
+     */
+    public CompletableFuture<Void> destroy(int partitionId, Function<T, CompletableFuture<Void>> destroyStorageFunction) {
+        DestroyStorageOperation destroyOp = (DestroyStorageOperation) operationByPartitionId.compute(partitionId, (partId, operation) -> {
+            checkStorageExists(partitionId);
+
+            throwExceptionDependingOnOperation(operation, partitionId);
+
+            return new DestroyStorageOperation();
+        });
+
+        return completedFuture(null)
```

Comment:
Why we need a `completedFuture` here? Can't we just use `destroyStorageFunction.apply(storageByPartitionId.getAndSet(partitionId, null))`

Reply:
This is a trick to make sure the operation gets rid of if it succeeds or fails.